PlayCover/ViewModel/StoreVM.swift

```diff
@@ -6,184 +6,192 @@
+        let semaphore = AsyncSemaphore(value: 0)
+        guard NetworkVM.isConnectedToNetwork() else { return }
+        if !sourcesList.isEmpty {
+            for index in sourcesList.indices {
+                sourcesList[index].status = .checking
+                let (sourceJson, sourceState) = await getSourceData(sourceLink: sourcesList[index].source)
+                sourcesList[index].status = sourceState
+                if sourceState == .valid {
+                    if let json = sourceJson {
+                        sourcesData.append(json)
+                        semaphore.signal()
+                    }
+                } else {
+                    semaphore.signal()
+                }
+                await semaphore.wait()
+            }
+        }
     }
 
-    func moveSourceUp(_ selected: inout Set<UUID>) {
-        let selectedData = self.sources.filter({ selected.contains($0.id) })
+    //
+    @discardableResult private func encode() -> Bool {
+        let encoder = PropertyListEncoder()
+        encoder.outputFormat = .xml
 
-        if let first = selectedData.first {
-            if var index = self.sources.firstIndex(of: first) {
-                index -= 1
-                self.sources.removeAll(where: { selected.contains($0.id) })
-                if index < 0 {
-                    index = 0
-                }
-                self.sources.insert(contentsOf: selectedData, at: index)
-            }
+        do {
+            let data = try encoder.encode(sourcesList)
+            try data.write(to: plistSource)
+            return true
+        } catch {
+            print("StoreVM: Failed to encode Sources.plist! ", error)
+            return false
         }
     }
 
-    func moveSourceDown(_ selected: inout Set<UUID>) {
-        let selectedData = self.sources.filter({ selected.contains($0.id) })
+    //
+    @discardableResult private func decode() -> Bool {
+        do {
+            let data = try Data(contentsOf: plistSource)
+            sourcesList = try PropertyListDecoder().decode([SourceData].self, from: data)
+            return true
+        } catch {
+            print("StoreVM: Failed to decode Sources.plist! ", error)
+            return false
+        }
+    }
 
-        if let first = selectedData.first {
-            if var index = self.sources.firstIndex(of: first) {
-                index += 1
-                self.sources.removeAll(where: { selected.contains($0.id) })
-                if index > self.sources.endIndex {
-                    index = self.sources.endIndex
+    //
+    private func getSourceData(sourceLink: String) async -> (SourceJSON?, SourceValidation) {
+        guard let url = URL(string: sourceLink) else { return (nil, .badurl) }
+        do {
+            let (data, response) = try await URLSession.shared.data(
+                for: URLRequest(url: url, cachePolicy: .reloadIgnoringLocalAndRemoteCacheData)
+            )
+            if !url.isFileURL {
```

Comment:
You should check for local files before starting any URLRequest

Reply:
Even if URL is local file we need to do URLRequest (or I am worng?)
So we are just checking afterwards to see if the url is local file and bypass the 200 statuscode check for it. nothing more

Reply:
As long it depends on response it does need to be checked, but there are better ways to check for local files, although the code works and the order is right, you should check for that before. just a suggestion.

Reply:
surely we need the `response` to be present to bypass it, right?

Reply:
looks like it was a misunderstanding. resolving this.